File: Project/NetworkingManaged/ClientSocket.cs
// Copyright 2019. All Rights Reserved.
using GameFramework.BinarySerializer;
using GameFramework.Common.Timing;
using System;
using System.Net;
using System.Net.Sockets;

namespace GameFramework.Networking
{
	public abstract class ClientSocket : BaseSocket
	{
		private abstract class ClientEventBase : EventBase
		{
		}

		private class ConnectedEvent : ClientEventBase
		{
		}

		private class ConnectionFailedEvent : ClientEventBase
		{
		}

		private class DisconnectedEvent : ClientEventBase
		{
		}

		private class BufferReceivedvent : ClientEventBase
		{
			public BufferStream Buffer
			{
				get;
				private set;
			}

			public BufferReceivedvent(BufferStream Buffer)
			{
				this.Buffer = Buffer;
			}
		}

		public delegate void ConnectionEventHandler();
		public delegate void BufferReceivedEventHandler(BufferStream Buffer);

		private double lastPingTime = 0;
		private double timeOffset = 0;

		public bool IsConnected
		{
			get;
			protected set;
		}

		public override bool IsReady
		{
			get { return SocketUtilities.GetIsReady(Socket); }
		}

		public override double Timestamp
		{
			get { return Time.CurrentEpochTime + timeOffset; }
		}

		public event ConnectionEventHandler OnConnected = null;
		public event ConnectionEventHandler OnConnectionFailed = null;
		public event ConnectionEventHandler OnDisconnected = null;
		public event BufferReceivedEventHandler OnBufferReceived = null;

		public ClientSocket(Protocols Type) : base(Type)
		{
		}

		public override void Service()
		{
			if (IsConnected && Statistics.LastTouchTime + Constants.PING_TIME <= Time.CurrentEpochTime)
			{
				Statistics.SetLastTouchTime(Time.CurrentEpochTime);

				SendPing();
			}

			base.Service();
		}

		public void Connect(string Host, ushort Port)
		{
			Connect(SocketUtilities.ResolveDomain(Host), Port);
		}

		public void Connect(IPAddress IP, ushort Port)
		{
			Connect(new IPEndPoint(IP, Port));
		}

		public void Connect(IPEndPoint EndPoint)
		{
			if (EndPoint.AddressFamily == AddressFamily.InterNetwork)
				EndPoint.Address = SocketUtilities.MapIPv4ToIPv6(EndPoint.Address);

			ConnectInternal(EndPoint);
		}

		public void Disconnect()
		{
			Shutdown();
		}

		protected abstract void ConnectInternal(IPEndPoint EndPoint);

		protected override void Receive()
		{
			try
			{
				int size = 0;

				lock (Socket)
				{
					if (Socket.Available == 0)
					{
						if (!IsReady)
							HandleDisconnection(Socket);

						return;
					}

					size = Socket.Receive(ReceiveBuffer);
				}

				Statistics.AddBandwidthIn((uint)size);

				uint index = 0;
				while (index < size)
				{
					uint packetSize = BitConverter.ToUInt32(ReceiveBuffer, (int)index);

					index += Packet.PACKET_SIZE_SIZE;

					HandleIncomingBuffer(new BufferStream(ReceiveBuffer, index, packetSize));

					index += packetSize;
				}
			}
			catch (SocketException e)
			{
				if (e.SocketErrorCode == SocketError.ConnectionReset)
				{
					HandleDisconnection(Socket);

					return;
				}

				throw e;
			}
		}

		protected abstract void HandleIncomingBuffer(BufferStream Buffer);

		protected override bool HandleSendCommand(SendCommand Command)
		{
			if (!SocketUtilities.GetIsReady(Socket))
				return false;

			SendOverSocket(Socket, Command.Buffer);

			return true;
		}

		protected override void ProcessEvent(EventBase Event)
		{
			ClientEventBase ev = (ClientEventBase)Event;

			if (ev is ConnectedEvent)
			{
				if (OnConnected != null)
					CallbackUtilities.InvokeCallback(OnConnected.Invoke);
			}
			else if (ev is ConnectionFailedEvent)
			{
				if (OnConnectionFailed != null)
					CallbackUtilities.InvokeCallback(OnConnectionFailed.Invoke);
			}
			else if (ev is BufferReceivedvent)
			{
				if (OnConnectionFailed != null)
					CallbackUtilities.InvokeCallback(OnBufferReceived.Invoke, ((BufferReceivedvent)ev).Buffer);
			}
		}

		protected override void HandleDisconnection(Socket Socket)
		{
			base.HandleDisconnection(Socket);

			if (MultithreadedCallbacks)
			{
				if (OnDisconnected != null)
					CallbackUtilities.InvokeCallback(OnDisconnected.Invoke);
			}
			else
			{
				AddEvent(new DisconnectedEvent());
			}

			IsConnected = false;
		}

		protected abstract void ProcessReceivedBuffer(BufferStream Buffer);

		protected void HandleReceivedBuffer(BufferStream Buffer)
		{
			if (MultithreadedCallbacks)
			{
				if (OnBufferReceived != null)
					CallbackUtilities.InvokeCallback(OnBufferReceived.Invoke, Buffer);
			}
			else
			{
				AddEvent(new BufferReceivedvent(Buffer));
			}
		}

		protected void RaiseOnConnectedEvent()
		{
			if (MultithreadedCallbacks)
			{
				if (OnConnected != null)
					CallbackUtilities.InvokeCallback(OnConnected.Invoke);
			}
			else
			{
				AddEvent(new ConnectedEvent());
			}
		}

		protected void RaiseOnConnectionFailedEvent()
		{
			if (MultithreadedCallbacks)
			{
				if (OnConnectionFailed != null)
					CallbackUtilities.InvokeCallback(OnConnectionFailed.Invoke);
			}
			else
			{
				AddEvent(new ConnectionFailedEvent());
			}
		}

		protected void HandlePingPacket(BufferStream Buffer)
		{
			double time = Time.CurrentEpochTime;

			double sendTime = Buffer.ReadFloat64();

			Statistics.SetLatency((uint)((time - sendTime) * 1000));

			double t0 = lastPingTime;
			double t1 = sendTime;
			double t2 = sendTime;
			double t3 = time;

			timeOffset = ((t1 - t0) + (t2 - t3)) / 2;
		}

		protected abstract BufferStream GetPingPacket();

		private void SendPing()
		{
			BufferStream pingBuffer = Packet.CreatePingBufferStream();

			lastPingTime = Time.CurrentEpochTime;

			SendOverSocket(Socket, pingBuffer);
		}
	}
}

File: Project/NetworkingNative/Compile/ClientSocket.cpp
// Copyright 2019. All Rights Reserved.
#include "..\Include\ClientSocket.h"
#include "..\Include\Constants.h"
#include "..\Include\Packet.h"
#include "..\Include\CallbackUtilities.h"
#include <Timing\Time.h>
#include <Utilities\TypeHelper.h>

using namespace GameFramework::Common::Timing;
using namespace GameFramework::Common::Utilities;

namespace GameFramework::Networking
{
	ClientSocket::ClientSocket(PlatformNetwork::IPProtocols Type) :
		BaseSocket(Type),
		m_IsConnected(false),
		m_LastPingTime(0),
		m_TimeOffset(0)
	{
	}

	void ClientSocket::Service(void)
	{
		if (m_IsConnected && GetStatistics().GetLastTouchTime() + Constants::PING_TIME <= Time::GetCurrentEpochTime())
		{
			GetStatistics().SetLastTouchTime(Time::GetCurrentEpochTime());

			SendPing();
		}

		BaseSocket::Service();
	}

	void ClientSocket::Connect(const std::string& Host, uint16_t Port)
	{
		Connect(SocketUtilities::ResolveDomain(Host), Port);
	}

	void ClientSocket::Connect(const IPAddress& IP, uint16_t Port)
	{
		Connect(IPEndPoint(IP, Port));
	}

	void ClientSocket::Connect(const IPEndPoint& EndPoint)
	{
		m_RemoteEndPoint = EndPoint;

		if (m_RemoteEndPoint.GetAddress().GetAddressFamily() == PlatformNetwork::AddressFamilies::InterNetwork)
			m_RemoteEndPoint.SetAddress(SocketUtilities::MapIPv4ToIPv6(m_RemoteEndPoint.GetAddress()));

		ConnectInternal(m_RemoteEndPoint);
	}

	void ClientSocket::Disconnect(void)
	{
		Shutdown();
	}

	void ClientSocket::Receive(void)
	{
		Socket socket = GetSocket();

		try
		{
			int availableSize = SocketUtilities::GetAvailableBytes(socket);

			if (availableSize == 0)
			{
				if (!GetIsReady())
					HandleDisconnection(socket);

				return;
			}

			std::byte* receiveBuffer = GetReceiveBuffer();

			uint32_t receiveBufferIndex = GetReceiveBufferIndex();

			uint32_t receiveSize = availableSize;
			if (!SocketUtilities::Receive(socket, receiveBuffer, receiveBufferIndex, receiveSize))
				return;

			GetStatistics().AddBandwidthIn(receiveSize);

			uint32_t totalSize = receiveBufferIndex + receiveSize;

			uint32_t index = 0;
			while (index != totalSize)
			{
				uint32_t packetSize = *(reinterpret_cast<uint32_t*>(receiveBuffer + index));

				if (totalSize < packetSize)
				{
					SetReceiveBufferIndex(totalSize);
					return;
				}

				index += Packet::PACKET_SIZE_SIZE;

				BufferStream buffer = BufferStream(receiveBuffer, index, packetSize);

				HandleIncomingBuffer(buffer);

				index += packetSize;
			}

			SetReceiveBufferIndex(0);
		}
		catch (PlatformNetwork::SocketException e)
		{
			if (e.GetError() == PlatformNetwork::Errors::ConnectionReset)
			{
				HandleDisconnection(socket);

				return;
			}

			throw e;
		}
	}

	bool ClientSocket::HandleSendCommand(SendCommand* Command)
	{
		if (!GetIsReady())
			return false;

		BaseSocket::SendOverSocket(GetSocket(), const_cast<BufferStream&>(Command->GetBuffer()));

		return true;
	}

	void ClientSocket::ProcessEvent(EventBase* Event)
	{
		ClientEventBase* ev = dynamic_cast<ClientEventBase*>(Event);

		if (IS_TYPE_OF(ev, ConnectedEvent))
		{
			CallbackUtilities::InvokeCallback(OnConnected);
		}
		else if (IS_TYPE_OF(ev, ConnectionFailedEvent))
		{
			CallbackUtilities::InvokeCallback(OnConnectionFailed);
		}
		else if (IS_TYPE_OF(ev, DisconnectedEvent))
		{
			CallbackUtilities::InvokeCallback(OnDisconnected);
		}
		else if (IS_TYPE_OF(ev, BufferReceivedvent))
		{
			CallbackUtilities::InvokeCallback(OnBufferReceived, dynamic_cast<BufferReceivedvent*>(Event)->GetBuffer());
		}
	}

	void ClientSocket::HandleDisconnection(Socket Socket)
	{
		BaseSocket::HandleDisconnection(Socket);

		if (GetMultithreadedCallbacks())
		{
			CallbackUtilities::InvokeCallback(OnDisconnected);
		}
		else
		{
			AddEvent(new DisconnectedEvent());
		}

		m_IsConnected = false;
	}

	void ClientSocket::HandleReceivedBuffer(const BufferStream& Buffer)
	{
		if (GetMultithreadedCallbacks())
		{
			CallbackUtilities::InvokeCallback(OnBufferReceived, Buffer);
		}
		else
		{
			AddEvent(new BufferReceivedvent(Buffer));
		}
	}

	double ClientSocket::GetTimestamp(void) const
	{
		return Time::GetCurrentEpochTime() + m_TimeOffset;
	}

	void ClientSocket::RaiseOnConnectedEvent(void)
	{
		if (GetMultithreadedCallbacks())
		{
			CallbackUtilities::InvokeCallback(OnConnected);
		}
		else
		{
			AddEvent(new ConnectedEvent());
		}
	}

	void ClientSocket::RaiseOnConnectionFailedEvent(void)
	{
		if (GetMultithreadedCallbacks())
		{
			CallbackUtilities::InvokeCallback(OnConnectionFailed);
		}
		else
		{
			AddEvent(new ConnectionFailedEvent());
		}
	}

	void ClientSocket::HandlePingPacket(BufferStream& Buffer)
	{
		double time = Time::GetCurrentEpochTime();

		double sendTime = Buffer.ReadFloat64();

		GetStatistics().SetLatency((time - sendTime) * 1000);

		double t0 = m_LastPingTime;
		double t1 = sendTime;
		double t2 = sendTime;
		double t3 = time;

		m_TimeOffset = ((t1 - t0) + (t2 - t3)) / 2;

		uint32_t payloadSize = Buffer.GetSize() - Packet::PING_SIZE;
		if (payloadSize != 0)
		{
			BufferStream buffer(Buffer.GetBuffer(), Packet::PING_SIZE - Packet::PACKET_SIZE_SIZE, payloadSize);
			HandlePingPacketPayload(buffer);
		}
	}

	void ClientSocket::SendPing(void)
	{
		BufferStream pingBuffer = GetPingPacket();

		m_LastPingTime = Time::GetCurrentEpochTime();

		BaseSocket::SendOverSocket(GetSocket(), pingBuffer);
	}
}